src/Persistence/Search/Query/HasWbFacetFeature.php

```diff
@@ -20,17 +28,46 @@ protected function getKeywords(): array {
 		return [ 'haswbfacet' ];
 	}
 
-	/**
-	 * @return array
-	 */
 	protected function doApply( SearchContext $context, $key, $value, $quotedValue, $negated ): array {
-		// TODO
+		$itemTypes = $this->getItemTypes( $context->getOriginalSearchTerm() );
+
+		if ( $itemTypes === [] ) {
+			return [ null, false ];
+		}
+
+		$constraints = $this->getConstraintsForValue( $value );
+
+		if ( $constraints === null ) {
+			return [ $this->itemTypeQueryBuilder->buildQuery( $itemTypes ), false ];
+		}
```

Comment:
This isn't very nice. We need to parse the full query string to get the current item type(s). But then we need to reparse just the current token to get the OR values, or a single AND value, since the actual query AND handling is done by the search engine (where each token's query gets AND-ed).

Reply:
Is this "just" a case of being wasteful by parsing too much?

Structurally the PR seems good. Some wasteful parsing probably has little or no real-world impact